fxa-oauth-server/lib/routes/authorization.js

```diff
@@ -218,104 +128,97 @@ module.exports = {
+    if (! client) {
+      logger.debug('notFound', { id: req.payload.client_id });
+      throw AppError.unknownClient(req.payload.client_id);
+    }
+    validateClientDetails(client, req.payload);
+
+    const grant = await validateRequestedGrant(claims, client, req.payload);
+    switch (req.payload.response_type) {
+    case RESPONSE_TYPE_CODE:
+      return await generateAuthorizationCode(client, req.payload, grant);
+    case RESPONSE_TYPE_TOKEN:
+      return await generateImplicitGrant(client, req.payload, grant);
+    default:
+      // Joi validation means this should never happen.
+      logger.critical('joi.response_type', { response_type: req.payload.response_type });
+      throw AppError.invalidResponseType();
+    }
+  }
+};
 
-        var uri = req.payload.redirect_uri || client.redirectUri;
 
-        if (uri !== client.redirectUri) {
-          logger.debug('redirect.mismatch', {
-            param: uri,
-            registered: client.redirectUri
-          });
+async function generateAuthorizationCode(client, payload, grant) {
+  // Clients must use PKCE if and only if they are a pubic client.
+  if (client.publicClient) {
+    if (! payload.code_challenge_method || ! payload.code_challenge) {
+      logger.info('client.missingPkceParameters');
+      throw AppError.missingPkceParameters();
+    }
+  } else {
+    if (payload.code_challenge_method || payload.code_challenge) {
+      logger.info('client.notPublicClient');
+      throw AppError.notPublicClient({ id: payload.client_id });
+    }
+  }
 
-          if (config.get('localRedirects') && isLocalHost(uri)) {
-            logger.debug('redirect.local', { uri: uri });
-          } else {
-            throw AppError.incorrectRedirect(uri);
-          }
+  const state = payload.state;
 
-        }
+  let code = await db.generateCode(Object.assign(grant, {
+    codeChallengeMethod: payload.code_challenge_method,
+    codeChallenge: payload.code_challenge,
+  }));
+  code = hex(code);
 
-        if (wantsGrant && ! client.canGrant) {
-          logger.warn('implicitGrant.notAllowed', {
-            id: req.payload.client_id
-          });
-          throw AppError.invalidResponseType();
-        }
+  const redirect = URI(payload.redirect_uri).addQuery({ code, state });
 
-        req.payload.redirect_uri = uri;
+  return {
+    code,
+    state,
+    redirect: String(redirect)
+  };
+}
 
-        checkPKCEParams(req, client);
+// N.B. We do not correctly implement the "implicit grant" flow from
+// RFC6749 which defines `response_type=token`. Instead we have a
+// privileged set of clients that use `response_type=token` for something
+// approximating the "resource owner password grant" flow, using an assertion
+// to just directly grant tokens for their own use.
+//
+// New clients should not do this, and should instead of `grant_type=fxa-assertion`
+// on the /token endpoint.
+//
+// This route is kept for backwards-compatibility only.
```

Comment:
TODO: note known users of this endpoint here, basically all current browser clients.